src/main/java/jenkins/scm/api/SCMHeadMigration.java

```diff
@@ -0,0 +1,126 @@
+package jenkins.scm.api;
+
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+
+/**
+ * If a {@link SCMSource} plugin needs to migrate the implementation classes for {@link SCMHead} this extension
+ * point allows the plugin to register type migrations. For speed of migration implementations should just jump direct
+ * to the final end-point and not expect recursive chain walking.
```

Comment:
well it also prevents having to worry about somebody accidentally setting up a circular migration chain via extensions